src/multiformats/address.cljc

```diff
@@ -2,29 +2,232 @@
+      (clojure.lang
+       Seqable
+       IObj
+       IMeta
+       IPersistentCollection))))
+
+(def protocol->attrs
+  "Map from address protocols to attributes used in encoding/decoding
+
+  https://github.com/multiformats/multiaddr/blob/master/protocols.csv"
+  {:ip4                {:code 0x04   :codec codec/ip4-codec}
+   :tcp                {:code 0x06   :codec codec/ushort-codec}
+   :udp                {:code 0x0111 :codec codec/ushort-codec}
+   :dccp               {:code 0x21   :codec codec/ushort-codec}
+   :ip6                {:code 0x21   :codec codec/ip6-codec}
+   :quic               {:code 0x01CC :codec codec/no-value-codec}
+   :udt                {:code 0x012D :codec codec/no-value-codec}
+   :utp                {:code 0x012E :codec codec/no-value-codec}
+   :ws                 {:code 0x01DD :codec codec/no-value-codec}
+   :wss                {:code 0x01DE :codec codec/no-value-codec}
+   :p2p-websocket-star {:code 0x1DF  :codec codec/no-value-codec}
+   :p2p-webrtc-star    {:code 0x113  :codec codec/no-value-codec}
+   :p2p-webrtc-direct  {:code 0x114  :codec codec/no-value-codec}
+   :p2p-circuit        {:code 0x0122 :codec codec/no-value-codec}
+   :dns                {:code 0x35   :codec codec/utf8-codec}
+   :dns4               {:code 0x36   :codec codec/utf8-codec}
+   :dns6               {:code 0x37   :codec codec/utf8-codec}
+   :dnsaddr            {:code 0x38   :codec codec/utf8-codec}})
+
+
+(def protocol->code
+  (into {} (map (juxt key (comp :code val)) protocol->attrs)))
+
+(def code->protocol
+  (into {} (map (juxt val key) protocol->code)))
+
+(defn- ensure-protocol-attrs
+  "look up protocol attributes, throws if unknown"
+  [protocol-key]
+  (let [protocol-attrs (get protocol->attrs protocol-key)]
+    (when-not protocol-attrs
+      (throw (ex-info (str "No protocol associated with " protocol-key)
+                      {:protocol protocol-key})))
+    protocol-attrs))
+
+(defn- copy-slice
+  ([src offset len]
+   (let [dst (b/byte-array len)]
+     (b/copy src offset dst 0 len)
+     dst))
+  ([src offset]
+   (copy-slice src offset (- (alength ^bytes src) offset))))
+
+
+(defn- decode-value
+  "decode value using protocol `codec` in `data` starting at `offset`"
+  [codec ^bytes data offset]
+  (if-let [fixed-len (codec/fixed-byte-length codec)]
+    (let [value-bytes (copy-slice data offset fixed-len)]
+      [(codec/bytes->str codec value-bytes) fixed-len])
+    ;; varint encoding
+    (let [[value-len num-read] (varint/read-bytes data offset)
+          value-bytes (copy-slice data (+ offset num-read) value-len)]
+      [(codec/bytes->str codec value-bytes) (+ num-read value-len)])))
+
+
+(defn- protocol-value-seq
+  "lazy sequence of `[protocol-key value]`, where
+  `value` is a string representation of the address protocol
+   value of `nil` if protocol has no value"
+  ([^bytes data]
+   (protocol-value-seq data 0))
+  ([^bytes data offset]
+   (when (< offset (alength data))
+       (lazy-seq
```

Comment:
Indent is a bit off here, but I'll probably be running `cljfmt` over this anyway.

Reply:
Ran `lein cljfmt fix` as its own commit.